p2p/src/peer_manager/peerdb/mod.rs

```diff
@@ -196,38 +195,77 @@ impl<S: PeerDbStorage> PeerDb<S> {
         &self,
         cur_outbound_conn_addr_groups: &BTreeSet<AddressGroup>,
         count: usize,
+    ) -> Vec<SocketAddress> {
+        self.select_non_reserved_outbound_addresses_with_rng(
+            cur_outbound_conn_addr_groups,
+            count,
+            &mut make_pseudo_rng(),
+        )
+    }
+
+    fn select_non_reserved_outbound_addresses_with_rng(
+        &self,
+        cur_outbound_conn_addr_groups: &BTreeSet<AddressGroup>,
+        count: usize,
+        rng: &mut impl Rng,
     ) -> Vec<SocketAddress> {
         if count == 0 {
             return Vec::new();
         }
 
         let now = self.time_getter.get_time();
 
-        // TODO: select new vs tried addresses with equal probability (or at least with a specific
-        // probability that is not directly tied to the sizes of both tables)
-        let mut selected = self
-            .addresses
-            .iter()
-            .filter_map(|(addr, address_data)| {
-                if address_data.connect_now(now)
+        let filter = |addr: &&SocketAddress| match self.addresses.get(addr) {
+            Some(addr_data) => {
+                addr_data.connect_now(now)
                     && !cur_outbound_conn_addr_groups
                         .contains(&AddressGroup::from_peer_address(&addr.as_peer_address()))
-                    && !address_data.reserved()
+                    && !addr_data.reserved()
                     && !self.banned_addresses.contains_key(&addr.as_bannable())
-                {
-                    Some(*addr)
-                } else {
-                    None
+            }
+            None => {
+                debug_assert!(false, "Address {addr} not found in self.addresses");
+                false
+            }
+        };
+
+        let mut selected_new =
```

Comment:
I think it's worth leaving a comment explaining why exactly the algorithm is like that.